File: applied-algorithms/4thought.py
# accepted by Kattis

m = int(input())
operations = ['+', '-', '*', '//']

possibleResult = {}

for op1 in operations:
    for op2 in operations:
        for op3 in operations:
            equationStr = '4 ' + op1 + ' 4 ' + op2 + ' 4 ' + op3 + ' 4'
            result = eval(equationStr)
            possibleResult[result] = equationStr.replace('//','/')

for i in range(m):
    n = int(input())
    if n in possibleResult:
        print(str(possibleResult[n]) + ' = ' + str(n))
    else:
        print('no solution')

File: applied-algorithms/4thought.js
// works but not accepted by all Kattis tests

var readline = require('readline');

var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const ops = ['+','-','*','/']

var lines = []; 

rl.on('line', function (line) {
    lines.push(line)
    
    let num = lines[0];
    
    if (lines[num]) {
        for (i=1; i<lines.length; i++) {
            fourThought(lines[i])
        }
        rl.close();
        process.exit(0);
    }
})

function fourThought(line) {
    for (const op1 in ops ) {
        for (const op2 in ops) {
            for (const op3 in ops) {
                let opString = '4 ' + ops[op1] + ' 4 ' + ops[op2] + ' 4 ' + ops[op3] + ' 4'
                
                if (eval(opString) == parseInt(line)){
                    console.log(opString, ' = ', line)
                    return
                }
            }
        }
    }
    console.log("no solution")
    return
}